pkg/container/ifset/set.go

```diff
@@ -0,0 +1,378 @@
+	switch v := member.(type) {
+	case Set:
+		// Remove members of the set instead the set itself
+		return s.removeAll(v.members)
+	case *Set:
+		// Remove members of the set instead the set itself
+		return s.removeAll(v.members)
+	default:
+		return s.removeAll(member)
+	}
+}
+
+func (s *Set) removeAll(member Member) (found, changed bool) {
+	if members, ok := member.(memberMap); ok {
+		// remove members in the map
+		for member := range members {
+			var f bool
+			f, changed = s.remove(member)
+			if f {
+				found = f
+			}
+		}
+		return found, changed
+	}
+	return s.remove(member)
+}
+
+func (s *Set) remove(member Member) (found, changed bool) {
+	members, ok := s.members.(memberMap)
+	if !ok || members == nil {
+		if !ok {
+			// One non-nil element in the set
+			if member == s.members {
+				s.Clear()
+				return true, true // found, changed
+			}
+		} else {
+			// Sole 'nil' in the set
+			if member == nil {
+				s.Clear()
+				return true, true // found, changed
+			}
+		}
+	} else {
+		oldLen := len(members)
+		delete(members, member)
+		newLen := len(members)
+		if oldLen > newLen {
+			if newLen == 1 {
+				// get to the single value and place it in place of the map
+				for member := range members {
+					if member == nil {
+						s.members = nilMember
+					} else {
+						s.members = member
+					}
+					return true, true // found, changed from a map to a singular value
+				}
+			}
+			return true, false // found, remains a map
+		}
+	}
+	return false, false // not found, not changed
+}
+
+// Clear makes the set '*s' empty.
+func (s *Set) Clear() {
+	s.members = nil
+}
+
+// Equal returns 'true' if the receiver and argument sets are the same.
+func (s Set) Equal(o Set) bool {
+	// Maps cannot be compared
+	sMembers, sOk := s.members.(memberMap)
+	oMembers, oOk := o.members.(memberMap)
+
+	// One is a map while the other is not?
+	if sOk != oOk {
+		return false
+	}
+
+	// Same singular members?
+	if !sOk && s.members == o.members || // both empty or same non-map interface value
+		sOk && sMembers == nil && oMembers == nil { // same singular 'nil'
+		return true
+	}
+
+	// Is only one empty?
+	if s.members == nil || o.members == nil {
+		return false
+	}
+
+	// by this point both of them have to be same size memberMaps
+	if !sOk || len(sMembers) != len(oMembers) {
+		return false
+	}
+
+	// compare the elements of the maps
+	for member := range sMembers {
+		if _, ok := oMembers[member]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// Members returns an iterator for the members in the set.
+func (s Set) Members() iter.Seq[Member] {
+	return func(yield func(m Member) bool) {
+		if members, ok := s.members.(memberMap); ok {
+			for member := range members {
+				if !yield(member) {
+					return
+				}
+			}
+		} else {
+			yield(s.members)
+		}
+	}
+}
+
+// ForEachOfType return an iterator for each member of type T in the set.
+func ForEachOfType[T any](s Set) iter.Seq[T] {
```

Comment:
AFAIU we need this data structure to hold heterogeneous types becase a `MapStateOwner` can be a`policyTypes.Key` or a `policy.CachedSelector`. Moreover we need it to be nullable.

Is that correct or am I missing any other types?

My hunch is that we are trying to deal with some tech debt on the policy side (namely in the way we are representing the owner of a map state entry) with a very specialized data structure for which I see very little opportunity to be reused.

Would it be possible to declare a single interface that describes all the possible types of owners of a map state entry? If so, we could use it to instantiate a `Set` like this one for the map state `owners`:

```go
type Set[T comparable] struct {
   ...
}
```

What the cost of that refactoring would be on the policy code?

Compared to the current implementation,I think  there will be multiple benefits with a Set like this one (broader reusability of the Set, easier implementation, better help from the compiler due to the type parameter, and so on).

Reply:
I had this first in `pkg/policy/types`, but thought maybe there is a chance that we need sets of heterogeneous interfaces elsewhere as well, so moved this to `pkg/containers/ifset` for this PR.

I'll see how this would work out using generics instead like you propose here.

Reply:
@pippolo84 Done. Using generics it is not possible to use the same struct member for the single element of the type `T` and for the `map[T]`. But this also simplifies the implementation quite a bit!

Also ended by by referring to the singular element via a pointer, as then we do not need to add a separate (boolean) member to tell if the singular member is occupied or not.

In the `BenchmarkRegenerateCIDRDenyPolicyRules` this performs more or less the same as the previous version.

Since the member can now be any `comparable` type, I moved the implementation to `pkg/container/set`.